src/wifi_dpp.rs

```diff
@@ -0,0 +1,451 @@
+    pub fn wait_for_credentials(&self) -> Result<ClientConfiguration, EspError> {
+        let next_state = self.bootstrapped.dpp.wait_for_next_state();
+        self.handle_next_state(next_state)
+    }
+
+    /// Blocking wait for credentials, a timeout, or a terminal error.  If the timeout is
+    /// reached, `Err(None)` is returned.
+    pub fn wait_for_credentials_with_timeout(
+        &self,
+        timeout: Duration,
+    ) -> Result<ClientConfiguration, Option<EspError>> {
+        match self
+            .bootstrapped
+            .dpp
+            .wait_for_next_state_with_timeout(timeout)
+        {
+            None => {
+                self.stop_listen();
+                Err(None)
+            }
+            Some(state) => Ok(self.handle_next_state(state)?),
+        }
+    }
+
+    fn handle_next_state(&self, state: DppState) -> Result<ClientConfiguration, EspError> {
+        match state {
+            DppState::ConfigurationReceived(c) => Ok(c),
+            DppState::Fail(e) => Err(e),
+            DppState::Stopped => {
+                info!("Caller requested DPP stop listening!");
+                Err(EspError::from_infallible::<ESP_ERR_INVALID_STATE>())
+            }
+            other => {
+                self.stop_listen();
+                Err(unexpected_state(other))
+            }
+        }
+    }
+
+    /// Stop listening for credentials.  If any callers are actively blocked waiting for credentials
+    /// they will be notified with EspError(ESP_ERR_INVALID_STATE).  This method is not
+    /// necessary to call after [Self::wait_for_credentials] or
+    /// [Self::wait_for_credentials_with_timeout] returns as the esp_dpp API automatically
+    /// stops listening on success or failure.
+    pub fn stop_listen(&self) {
+        info!("Stopping DPP listener...");
+
+        // SAFETY: This function should be safe to call without any locks as it mostly just
+        // sets a flag to halt at next chance.
+        unsafe { esp_supp_dpp_stop_listen() };
+
+        self.bootstrapped.dpp.set_pending_state(DppState::Stopped);
+    }
+
+    /// Attempt to retry after a transient failure in [Self::wait_for_credentials] (for example
+    /// if a user scanned a bogus QR code or there was a recoverable transmit error on the
+    /// channel).  This API will fail at runtime if ESP-IDF is not patched with
+    /// https://github.com/espressif/esp-idf/pull/10865.
+    ///
+    /// On failure to retry, the caller is expected to bootstrap again in order to logically retry
+    /// however this will potentially lose state and generate a new QR code if any parameters
+    /// change (such as a new key being generated).
+    pub fn attempt_retry(self) -> Result<EspWifiDpp<'d, 'w, T>, ()> {
+        if Self::is_start_listen_patched() {
+            Ok(self.bootstrapped)
+        } else {
+            Err(())
+        }
+    }
+
+    // TODO: Sure would be nice to be able to write esp_idf_version >= 5.1...
+    #[cfg(any(
```

Comment:
Note: I could do this at runtime for much less code, but I'm not sure how to access the version from esp-idf-sys?  Is ESP_IDF_VERSION/_VAL exposed?